### PR TITLE
Update libsecp256k1 and use RFC6979 extra entropy for test cases

### PR DESCRIPTION
**Pull request by sipa**
Included libsecp256k1 changes (multiple authors):
- Converted some runtime initialization to static constants.
- Convert all code to strict C89 (+ long long, + __int128).
- Get rid of variable-length string conversion code.
- Introduce a compact storage type, and use it for precomputed tables.
- Extra benchmarks.
- More x86_64 assembly (speeds up signing mostly).
- Fast byteswapping inside hash functions.
- Make the built-in RFC6979 nonce generator support additional entropy.
- Use RFC6979 as test RNG.

### CODE REVIEW

**Comment by gmaxwell** [comment]
ACK.

I was uncomfortable with (and complained about) the "test nonce" approach used previously as it used seriously insecure linearly related nonces... this was fine for tests but it seemed like a matter of time before someone wanted to use it to derandomize regular signing (and, indeed, someone recently asked about that).

File: src/key.cpp [hunk 1]
```diff
@@ -73,25 +74,14 @@ CPubKey CKey::GetPubKey() const {
     return result;
 }
 
-extern "C"
```
Comment (laanwj) [inline src/key.cpp]: Happy to get rid of this function.

**Comment by laanwj** [comment]
ACK to bitcoin core changes, haven't been able to review all secp256k1 changes.

**Comment by sipa** [comment]
Would it be interesting to have a more automated way for verifying that the contents of the subtree directory matches the contents of the listed commit of the subtree repository?

**Comment by laanwj** [comment]
Yes, a script that compares our secp256k1 directory with an upstream repository would be useful.
Probably amounts to one, fairly complicated, `git diff` statement but my knowledge of git subtrees is not enough to come up with it.

**Comment by theuni** [comment]
upstream secp256k1 commit is 1897b8e

```git diff 1897b8e origin/pr/5952:src/secp256k1```

**Comment by sipa** [comment]
Another way is checking the git tree id:

(in master)
$ git ls-tree master src/secp256k1
040000 tree 3b675a638355f8ccad0474900fed0bffcda8e62a    src/secp256k1
(in secp256k1, with 50cc6ab the commit id of the last secp256k1 import)
$ git show --format=raw 50cc6ab | head -n2
commit 50cc6ab0625efda6dddf1dc86c1e2671f069b0d8
tree 3b675a638355f8ccad0474900fed0bffcda8e62a

**Comment by laanwj** [comment]
Checked the subtree using #5965's `git-subtree-check.sh` script.
```
src/secp256k1 in HEAD was last updated to upstream commit 1897b8e90bbbdcd919427c9a8ae35b420e919d8f (tree 1c8df9fa9ddfb2b035ac0327fe074b634e458f56)
src/secp256k1 in HEAD currently refers to tree 1c8df9fa9ddfb2b035ac0327fe074b634e458f56
GOOD
```
ACK

**Comment by SergioDemianLerner** [comment]
I did a first pass review of all changes, but the number of changes is huge! Most of them are simply the removal of a variable declaration from its first use. However, many reviewers of this commit would be helpful.

**Comment by gmaxwell** [comment]
@SergioDemianLerner This is a subtree merge from the libsecp256k1 repository, merging in months of work; you can see the individual changes there (and the review process there), it's much easier to review from there because e.g. with the changes split out you can verify various changes that shouldn't have changed the binaries didn't and such.

**Comment by sipa** [comment]
In addition, with the script added in #5965 you can verify that the subtree
corresponds to the result of the commits in the external repository.

**Pull request merged**